ceph/parallel.py

```diff
@@ -1,121 +1,95 @@
+        self._executor = ThreadPoolExecutor() if thread_pool else ProcessPoolExecutor()
+        self._timeout = timeout
+        self._futures = list()
+        self._results = list()
 
-    If one of the spawned functions throws an exception, it will be thrown
-    when iterating over the results, or when the with block ends.
+    def spawn(self, fun, *args, **kwargs):
+        """Triggers the first class method.
 
-    At the end of the with block, the main thread waits until all
-    spawned functions have completed, or, if one exited with an exception,
-    kills the rest and raises the exception.
-    """
+        Args:
+            func:       Function to be executed.
+            args:       A list of variables to be passed to the function.
+            kwargs      A dictionary of named variables.
 
-    def __init__(self):
-        self.group = gevent.pool.Group()
-        self.results = gevent.queue.Queue()
-        self.count = 0
-        self.any_spawned = False
-        self.iteration_stopped = False
-
-    def spawn(self, func, *args, **kwargs):
-        self.count += 1
-        self.any_spawned = True
-        greenlet = self.group.spawn(capture_traceback, func, *args, **kwargs)
-        greenlet.link(self._finish)
+        Returns:
+            None
+        """
+        _future = self._executor.submit(fun, *args, **kwargs)
+        self._futures.append(_future)
 
     def __enter__(self):
         return self
 
-    def __exit__(self, type_, value, traceback):
-        self.group.join()
+    def __exit__(self, exc_type, exc_value, trackback):
+        try:
+            for _f in as_completed(self._futures, timeout=self._timeout):
+                self._results.append(_f.result())
 
-        if value is not None:
             return False
+        except Exception as e:
+            logger.exception(e)
+            raise e
 
-        try:
-            # raises if any greenlets exited with an exception
-            for result in self:
-                log.debug("result is %s", repr(result))
-                pass
-        except Exception:
-            # Emit message here because traceback gets stomped when we re-raise
-            log.exception("Exception in parallel execution")
-            raise
         return True
 
     def __iter__(self):
         return self
 
     def __next__(self):
-        if not self.any_spawned or self.iteration_stopped:
-            raise StopIteration()
-        result = self.results.get()
-
-        try:
-            resurrect_traceback(result)
-        except StopIteration:
-            self.iteration_stopped = True
-            raise
-
-        return result
-
-    def _finish(self, greenlet):
-        if greenlet.successful():
-            self.results.put(greenlet.value)
-        else:
-            self.results.put(greenlet.exception)
-
-        self.count -= 1
-        if self.count <= 0:
-            self.results.put(StopIteration())
+        for r in self._results:
+            yield r
```

Comment:
I suggest one more method for closing all threads when exit status runs bad with custom option to set that flag.

Reply:
Thanks @sunilkumarn417 

It is a good suggestion. Let me add it.

Reply:
Done. There are two options that are natively supported. The latest commit exposes those options to the end user.

This way when an exception is encountered, we can control the behavior.